### ISSUE TITLE
t/cme-command.t fails with Config-Model-2.134

### ISSUE THREAD

**Issue by ppisar** [dod38fr/cme-perl#4]
t/cme-command.t test fails after upgrading Config-Model from 2.133 to 2.134:

~~~~
t/cme-command.t .. 1/?
    #   Failed test 'threw no exceptions'
    #   at t/cme-command.t line 257.
    #          got: 'Error in script wr_root/popcon1/my-script.cme: Attribute (start_node) does not pass the type constraint because: Validation failed for 'Config::Model::Node' with value undef at /usr/lib64/perl5/vendor_perl/Mouse/Util.pm line 395.
    #   Mouse::Util::throw_error(Mouse::Meta::Attribute=HASH(0x555d0a85b928), "Attribute (start_node) does not pass the type constraint beca"..., "data", undef, "depth", -1) called at /usr/share/perl5/vendor_perl/Config/Model/Instance.pm line 390
    #   Config::Model::Instance::load(Config::Model::Instance=HASH(0x555d0beb67f8), "step", "! MY_HOSTID=\$namefoobar3") called at /usr/share/perl5/vendor_perl/Config/Model/Instance.pm line 384
    #   Config::Model::Instance::modify(Config::Model::Instance=HASH(0x555d0beb67f8), "! MY_HOSTID=\$namefoobar3") called at (eval 257) line 3
    #   eval 'use Config::Model qw(cme);
    # my ($opt,$val,$name) = @ARGV;
    # cme(application => "popcon", root_dir => $val)->modify("! MY_HOSTID=\\$name$name");
    # ' called at /home/test/fedora/perl-App-Cme/App-Cme-1.029/lib/App/Cme/Command/run.pm line 121
    #   App::Cme::Command::run::execute(App::Cme::Command::run=HASH(0x555d0bec8490), Getopt::Long::Descriptive::Opts::__OPT__::32=HASH(0x555d0beecea8), ARRAY(0x555d0be9ded0)) called at /usr/share/perl5/vendor_perl/App/Cmd.pm line 468
    #   App::Cmd::execute_command(App::Cme=HASH(0x555d0bed1190), App::Cme::Command::run=HASH(0x555d0bec8490), Getopt::Long::Descriptive::Opts::__OPT__::32=HASH(0x555d0beecea8), "wr_root/popcon1/my-script.cme", "foobar3") called at /usr/share/perl5/vendor_perl/App/Cmd.pm line 321
    #   App::Cmd::run(App::Cme=HASH(0x555d0bed1190)) called at /usr/share/perl5/vendor_perl/App/Cmd/Tester.pm line 125
    #   eval {...} called at /usr/share/perl5/vendor_perl/App/Cmd/Tester.pm line 122
    #   App::Cmd::Tester::_run_with_capture("App::Cmd::Tester", App::Cme=HASH(0x555d0bed1190), ARRAY(0x555d0beb7230)) called at /usr/share/perl5/vendor_perl/App/Cmd/Tester.pm line 92
    #   App::Cmd::Tester::test_app("App::Cmd::Tester", "App::Cme", ARRAY(0x555d0beb7230)) called at /usr/share/perl5/vendor_perl/Sub/Exporter/Util.pm line 18
    #   Sub::Exporter::Util::__ANON__("App::Cme", ARRAY(0x555d0beb7230)) called at t/cme-command.t line 256
    #   main::__ANON__() called at /usr/share/perl5/vendor_perl/Test/Builder.pm line 334
    #   eval {...} called at /usr/share/perl5/vendor_perl/Test/Builder.pm line 334
    #   Test::Builder::subtest(Test::Builder=HASH(0x555d096f0408), "213: modification with a Perl script run by cme run with args", CODE(0x555d0ac33e70)) called at /usr/share/perl5/vendor_perl/Test/More.pm line 809
    #   Test::More::subtest("213: modification with a Perl script run by cme run with args", CODE(0x555d0ac33e70)) called at t/cme-command.t line 264
    #
    # '
    #     expected: undef

    #   Failed test 'all went well'
    #   at t/cme-command.t line 258.
    #          got: '-1'
    #     expected: '0'
    # Failed command: run wr_root/popcon1/my-script.cme -root-dir wr_root/popcon1 foobar3

    #   Failed test 'updated MY_HOSTID with script'
    #   at t/cme-command.t line 261.
    #     File wr_root/popcon1/etc/popularity-contest.conf contents do not match /(?^:"\$namefoobar3")/
    # Looks like you failed 3 tests of 5.

#   Failed test '213: modification with a Perl script run by cme run with args'
#   at t/cme-command.t line 264.

    #   Failed test 'run "223: modification with a script and var section which uses regexp and capture": stdout content'
    #   at t/cme-command.t line 263.
    #          got: '
    # Changes applied to popcon configuration:
    # - MY_HOSTID: '$namefoobar2' -> 'aaaaab'
    # - MY_HOSTID: 'aaaaab' -> 'aaaax4ab'
    #
    # '
    #     expected: '
    # Changes applied to popcon configuration:
    # - MY_HOSTID: '$namefoobar3' -> 'aaaaab'
    # - MY_HOSTID: 'aaaaab' -> 'aaaax4ab'
    #
    # '
    # Looks like you failed 1 test of 5.

#   Failed test '223: modification with a script and var section which uses regexp and capture'
#   at t/cme-command.t line 264.
# Looks like you failed 2 tests of 20.
~~~~

Git bisect on Config-Model identifies this commit as the culprit:

~~~~
commit 3390cad4f411557f2efa14faf8ddb503c36024e3 (HEAD, refs/bisect/bad)
Author: Dominique Dumont <dod@debian.org>
Date:   Sat Apr 20 09:05:37 2019 +0200

    Add Instance->erset_config method

diff --git a/lib/Config/Model/Instance.pm b/lib/Config/Model/Instance.pm
index 42c4c184..15a57a89 100644
--- a/lib/Config/Model/Instance.pm
+++ b/lib/Config/Model/Instance.pm
@@ -255,13 +255,21 @@ has [qw/config_dir config_file/] => (
 has tree => (
     is      => 'ro',
     isa     => 'Config::Model::Node',
-    builder => 'reset_config',
+    builder => '_build_tree',
+    lazy    => 1,
+    clearer => '_clear_config',
     reader  => 'config_root',
     handles => [qw/apply_fixes deep_check grab grab_value/],
 );

 sub reset_config {
     my $self = shift;
+    $self->_clear_config;
+    return $self->config_root;
+}
+
+sub _build_tree {
+    my $self = shift;

     return $self->load_node (
         config_class_name => $self->{root_class_name},
~~~~

**Comment by ppisar** [comment]
I isolated it to making Config::Model::Instance->tree lazy. If I remove the lazy  => 1, all tests pass.

**Comment by dod38fr** [comment]
Hello Petr

I'm really sorry to reply this long after your bug report. I guess the mail from github went below my radar.

I've stumbled on this issue in May. I believe this is fixed in Config::Model by https://github.com/dod38fr/config-model/commit/36f6a33fc2dc64698e8e1eef0177c06b1f8c1c0c which was released in Config::Model 2.135 at the beginning of June.

Thanks for taking the time to write a report with a `git bisect`

All the best
